src/com/facebook/buck/parser/UnexpectedFlavorException.java

```diff
@@ -105,22 +107,30 @@ public static UnexpectedFlavorException createWithSuggestions(
   }
 
   private static ImmutableSet<Flavor> getInvalidFlavors(Flavored flavored, BuildTarget target) {
-    ImmutableSet.Builder<Flavor> builder = ImmutableSet.builder();
-    for (Flavor flavor : target.getFlavors()) {
-      if (!flavored.hasFlavors(ImmutableSet.of(flavor))) {
-        builder.add(flavor);
-      }
-    }
-    return builder.build();
+    return target.getFlavors().stream()
+        .filter(flavor -> !flavored.hasFlavors(ImmutableSet.of(flavor)))
+        .collect(MoreCollectors.toImmutableSet());
+  }
+
+  private static ImmutableSet<Flavor> getValidFlavors(Flavored flavored, BuildTarget target) {
+    return target.getFlavors().stream()
+        .filter(flavor -> flavored.hasFlavors(ImmutableSet.of(flavor)))
+        .collect(MoreCollectors.toImmutableSet());
   }
 
   private static String createDefaultMessage(
-      BuildTarget target, ImmutableSet<Flavor> invalidFlavors) {
-    ImmutableSet<String> invalidFlavorsStr =
-        invalidFlavors.stream().map(Flavor::toString).collect(MoreCollectors.toImmutableSet());
+      BuildTarget target, ImmutableSet<Flavor> invalidFlavors, ImmutableSet<Flavor> validFlavors) {
+    String invalidFlavorsStr = invalidFlavors.stream()
+        .map(Flavor::toString)
+        .collect(Collectors.joining(", "));
+
+    String validFlavorsStr = validFlavors.stream()
+        .map(Flavor::getName)
+        .collect(Collectors.joining(", "));
+
     String invalidFlavorsDisplayStr = String.join(", ", invalidFlavorsStr);
     return String.format(
-        "The following flavor(s) are not supported on target %s:\n%s.",
-        target, invalidFlavorsDisplayStr);
+        "The following flavor(s) are not supported on target %s:\n%s. Available flavors are: %s",
```

Comment:
this line also got a following comment from one of our engineers:
```
In the test plan, there are lots of similar platforms. Maybe try to de-dedup as much as possible.
In all cases, if the number of items in the list is larger than 4, there should be one item per line.
```

Reply:
Hmm... What should I dedup, given both methods use Sets?

Reply:
I think comment author meant that `macosx-i386, macosx-x86_64` can be grouped into a single category, but I don't think it's a good thing to do, but writing flavors on separate lines makes sense to me.

Reply:
Well, that could be a separate feature - grouping flavors and platforms, but it also has an edge case: what if some flavors do not support some specific platforms?

Will split flavors by a newline and update this PR.